src/csharp/Grpc.Core/Channel.cs

```diff
@@ -299,6 +304,40 @@ private ChannelState GetConnectivityState(bool tryToConnect)
             }
         }
 
+        /// <summary>
+        /// Constantly Watches channel connectivity status to work around https://github.com/GoogleCloudPlatform/google-cloud-dotnet/issues/822
+        /// </summary>
+        private async Task RunConnectivityWatcherAsync()
+        {
+            try
+            {
+                var lastState = State;
+                while (lastState != ChannelState.Shutdown)
+                {
+                    lock (myLock)
+                    {
+                        if (shutdownRequested)
+                        {
+                            break;
+                        }
+                    }
+
+                    try
+                    {
+                        await WaitForStateChangedAsync(lastState, DateTime.UtcNow.AddSeconds(1)).ConfigureAwait(false);
```

Comment:
I left out an important note: constantly using a timeout on watch connectivity state calls like this could lead to steady memory accumulation over  the lifetimes of long-lived channel objects.

(right now for every timeout on that call, there's a malloc that won't get free'd until the channel is later destroyed).

see https://github.com/grpc/grpc/pull/10813 for a PR that mentions the issue and tries to fix it.

That PR is held up on an unresolved test flake that it seems to be causing; currently I'm revising the ruby state watches to use infinite timeouts, with `grpc_destroy_channel` ending ops by causing state change to `shutdown`, when process is shutdown.

That said I'll work more on getting https://github.com/grpc/grpc/pull/10813 in, but I'm not sure how quickly that can happen.